pkg/chaosdaemon/tasks/groupprocesshandle.go

```
@@ -0,0 +1,108 @@
+// limitations under the License.
+//
+
+package tasks
+
+import (
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/chaosdaemon/util"
+)
+
+type ChaosOnGroupProcess interface {
+	Fork() (ChaosOnGroupProcess, error)
+	AssignChaosOnProcess
+
+	ChaosOnProcess
+	ChaosCanRecover
+}
+
+// GroupProcessHandler implements Group PID inject & recover.
+type GroupProcessHandler struct {
+	Main     ChaosOnGroupProcess
+	childMap map[PID]ChaosOnGroupProcess
+	logger   logr.Logger
+}
+
+func NewGroupProcessHandler(logger logr.Logger, main ChaosOnGroupProcess) GroupProcessHandler {
+	return GroupProcessHandler{
+		Main:     main,
+		childMap: make(map[PID]ChaosOnGroupProcess),
+		logger:   logger,
+	}
+}
+
+func (gp *GroupProcessHandler) Inject(pid PID) error {
+	childPids, err := util.GetChildProcesses(uint32(pid), gp.logger)
+	if err != nil {
+		gp.logger.Error(err, "failed to get child process")
+	}
+
+	err = gp.Main.Inject(pid)
+	if err != nil {
+		return errors.Wrapf(err, "inject main process : %d", pid)
+	}
+
+	for _, childPid := range childPids {
+		childPID := PID(childPid)
+		if childProcessChaos, ok := gp.childMap[childPID]; ok {
+			err := gp.Main.Assign(childProcessChaos)
+			if err != nil {
+				return err
+			}
+			err = childProcessChaos.Inject(childPID)
+			if err != nil {
+				gp.logger.Error(err, "failed to inject old child process")
+			}
+		} else {
+			childProcessChaos, err := gp.Main.Fork()
+			if err != nil {
+				gp.logger.Error(err, "failed to create child process")
+				return nil
+			}
+			err = childProcessChaos.Inject(pid)
+			if err != nil {
+				gp.logger.Error(err, "failed to inject new child process")
+				return nil
+			}
+			gp.childMap[childPID] = childProcessChaos
+		}
+	}
+	return nil
+}
+
+func (gp *GroupProcessHandler) Recover(pid PID) error {
+	childPids, err := util.GetChildProcesses(uint32(pid), gp.logger)
+	if err != nil {
+		gp.logger.Error(err, "failed to get child process")
```

do we need to return the error?

---

I have to put it here because I think we must make sure the chaos on main PID can be recover.

---

And `GetChildProcesses` only return error when `read /proc/pid/tasks  error, ppid : %d`. We should try to recover main process here.